File: java/LongestSubstring.java
public class LongestSubstring {
    public static int lengthOfLongestSubstring(String s) {
        int[] lastIndex = new int[128]; // ASCII
        for (int i = 0; i < 128; i++) lastIndex[i] = -1;

        int maxLen = 0, start = 0;
        for (int end = 0; end < s.length(); end++) {
            char c = s.charAt(end);
            if (lastIndex[c] >= start) {
                start = lastIndex[c] + 1;
            }
            lastIndex[c] = end;
            maxLen = Math.max(maxLen, end - start + 1);
        }
        return maxLen;
    }

    public static void main(String[] args) {
        System.out.println(lengthOfLongestSubstring("abcabcbb")); // Output: 3
    }
}

File: go/longest_substring.go
#find the longest substring without repeating characters
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
    lastIndex := make(map[rune]int)
    start, maxLen := 0, 0

    for i, ch := range s {
        if idx, ok := lastIndex[ch]; ok && idx >= start {
            start = idx + 1
        }
        lastIndex[ch] = i
        if i-start+1 > maxLen {
            maxLen = i - start + 1
        }
    }
    return maxLen
}

func main() {
    s := "abcabcbb"
    fmt.Println("Longest substring length:", lengthOfLongestSubstring(s))
}